### PR TITLE
Improve MIME boundary parameter handling

### PR DESCRIPTION
**Pull request by val-ms**
## Summary

This is a follow-up to #1720 and improves email attachment extraction for six
additional MIME boundary parameter forms reported by the same researcher.

The mail parser now resolves a canonical multipart boundary from the complete
`Content-Type` value while preserving the existing general MIME argument
parser. This covers:

- leading whitespace in a boundary value;
- malformed unquoted values containing spaces;
- quoted-pair backslash and escaped-quote handling; and
- RFC 2231 boundary continuations, including encoded segments.

Six encrypted `.eml` fixtures reconstructed from the report verify that the
standard ClamAV test attachment is extracted and detected in each case. As
with the earlier email-evasion work, this is a parser compatibility and
detection-hardening improvement rather than a product vulnerability.

## Testing

- `make -j12`
- full `ctest -V` suite (all seven targets passed)
- `make install`
- `git diff --check`
- pre-fix and post-fix scans of all six reconstructed messages

Reported by Artem Danilov (Positive Technologies).

CLAM-3074

### CODE REVIEW

File: libclamav/mbox.c [hunk 1]
```diff
@@ -3637,6 +3918,36 @@ parseMimeHeader(message *m, const char *cmd, const table_t *rfc821Table, const c
...
+                    if (boundaryArgument == NULL) {
+                        free(contentTypeBoundary);
+                        if (copy)
+                            free(copy);
+                        free(buf);
+                        return PARSE_HEADER_ALLOC_FAIL;
+                    }
+
+                    snprintf(boundaryArgument, boundaryArgumentSize, "boundary=%s", contentTypeBoundary);
+                    messageAddArgument(m, boundaryArgument);
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Good catch—thank you. The single `boundary*=` forms are deduplicated before a second decode, but I reproduced both failure modes with RFC 2231 continuations, where the earlier partial argument has already failed normalization: a split UTF-8 boundary lost its high bits, and a split encoded `*=` sequence was reinterpreted as RFC 2231 syntax. Commit c68bd1d29 adds an already-decoded storage path that retains the existing argument validation, duplicate handling, ownership, and MIME-type behavior, plus encrypted end-to-end fixtures for both cases. Both messages now extract and detect the embedded ClamAV test file. The full local CTest suite and install also pass.

File: libclamav/mbox.c [hunk 2]
```diff
@@ -3637,6 +3918,36 @@ parseMimeHeader(message *m, const char *cmd, const table_t *rfc821Table, const c
...
+                    free(buf);
+                    return PARSE_HEADER_ALLOC_FAIL;
+                }
+
+                /*
+                 * messageAddArgument() treats a leading pair of quotes as syntax.
+                 * Leave values that would be reparsed differently on the existing path.
+                 */
+                if ((contentTypeBoundary != NULL) && (*contentTypeBoundary != '\0') &&
+                    ((*contentTypeBoundary != '"') || (strchr(contentTypeBoundary + 1, '"') == NULL))) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in 02bcf7b53. Canonical boundary lookup now uses a boundary-specific literal-value path, so already-decoded leading and embedded quote characters are preserved rather than interpreted as quoting syntax. I added the encrypted end-to-end fixture `clam.mail-boundary-rfc2231-literal-quotes.eml.xor`, which extracts and detects the embedded test file. The full local CTest suite passes.

File: libclamav/mbox.c [hunk 3]
```diff
@@ -3472,13 +3473,292 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+        while ((nameEnd > nameStart) && isspace((unsigned char)nameEnd[-1]))
+            nameEnd--;
+
+        nameLength = (size_t)(nameEnd - nameStart);
+        if ((nameLength < 8) || (strncasecmp(nameStart, "boundary", 8) != 0))
+            continue;
+
+        suffix = nameStart + 8;
+        if (nameLength == 8) {
+            if (decodeMimeBoundaryValue(separator + 1, false, false, &decoded) < 0) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed this was a regression and fixed it in 02bcf7b53. `nextMimeArgument()` now treats backslash as an escape only inside quoted strings, so the unquoted semicolon splits the declarations and the later `boundary=Y` remains effective, while quoted-pair handling is preserved. I added the encrypted end-to-end fixture `clam.mail-boundary-unquoted-escaped-semicolon.eml.xor`; the full local CTest suite passes.

File: libclamav/mbox.c [hunk 4]
```diff
@@ -3497,6 +3778,7 @@ parseMimeHeader(message *m, const char *cmd, const table_t *rfc821Table, const c
     } else {
         ptr = arg;
     }
+    headerValue = ptr;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in bc7ec2d75. Canonical boundary parsing now removes RFC 822 comments while preserving quoted-pair backslashes, so `decodeMimeBoundaryValue()` consumes each escape exactly once. Existing `rfc822comments()` callers retain their prior decoding behavior. I added the encrypted end-to-end fixture `clam.mail-boundary-comment-quoted-pair.eml.xor`, with the detectable attachment under the intended `--\X` delimiter. The direct scan, focused libclamav test, and full 7-target CTest suite all pass.

File: libclamav/mbox.c [hunk 5]
```diff
@@ -3472,13 +3474,300 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+    size_t order                                            = 0;
+    size_t i;
+    int status = 0;
+
+    *boundary = NULL;
+
+    if (contentType == NULL)
+        return 0;
+
+    commentStrippedContentType = rfc822comments(contentType, NULL, true);
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3939]: Confirmed and fixed in e8c2fd636. Quotes inside RFC 822 comments are now ignored both by comment removal and by the earlier unbalanced-quote/header-folding check; the latter was another path where the same quote could consume following message lines. I added the encrypted end-to-end fixture `clam.mail-boundary-comment-quote.eml.xor`, which verifies that the embedded test file is extracted and detected. The direct scan, focused libclamav test, and full CTest suite all pass.

File: libclamav/mbox.c [hunk 6]
```diff
@@ -3472,13 +3474,300 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+
+        if (decodeMimeBoundaryValue(separator + 1, encoded, section == 0, &decoded) < 0) {
+            status = -1;
+            goto done;
+        }
+
+        free(segments[section]);
+        segments[section] = decoded;
+        if (section == 0)
+            continuationOrder = order;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:4047]: Confirmed and fixed in d9dedd7fd. A repeated RFC 2231 section zero now starts a fresh continuation and clears higher-numbered segments retained from the earlier instance. The first section zero deliberately preserves any segments that arrived before it, so the existing out-of-order tolerance remains intact. I added the encrypted end-to-end fixture `clam.mail-boundary-rfc2231-repeated-section-zero.eml.xor`; it failed as clean before the fix and now extracts and detects the embedded test file. I also directly verified the out-of-order form, and the focused libclamav test, analyzer pass, and full CTest suite pass.

File: libclamav/mbox.c [hunk 7]
```diff
@@ -3443,7 +3444,8 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
             } else {
                 switch (*p) {
                     case '\\':
-                        backslash = true;
+                        if (inquote)
+                            backslash = true;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3714]: Confirmed and fixed in 16f62b073. The tokenizer now preserves its legacy unquoted quoted-pair behavior for ordinary Content-Type and Content-Disposition parsing; the delimiter behavior needed by canonical boundary selection is enabled only for that boundary-specific call. I added the encrypted fixture `clam.mail-content-disposition-escaped-semicolon.eml.xor` and a clamscan metadata test that verifies the complete `foo\;bar.exe` filename survives parsing. The existing escaped-semicolon boundary fixture still passes, as do the focused tests, analyzer pass, and full CTest suite.

File: libclamav/mbox.c [hunk 8]
```diff
@@ -3472,13 +3484,309 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+
+    next         = (commentStrippedContentType != NULL) ? commentStrippedContentType : contentType;
+    argumentSize = strlen(next) + 1;
+    argument     = cli_max_malloc(argumentSize);
+    if (argument == NULL) {
+        free(commentStrippedContentType);
+        return -1;
+    }
+
+    while ((next = nextMimeArgument(next, argument, argumentSize, false)) != NULL) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in d51bc598f. Canonical parsing now treats an escaped semicolon as a delimiter only when its look-ahead identifies another boundary parameter (including RFC 2231 forms). A lone escaped semicolon remains part of the existing boundary value. I added the encrypted end-to-end fixture `clam.mail-boundary-lone-escaped-semicolon.eml.xor`; it was clean before this fix and now extracts and detects the embedded test file. I also directly rechecked the later-boundary and Content-Disposition filename cases. The analyzer pass, focused tests, and full CTest suite all pass.

File: libclamav/mbox.c [hunk 9]
```diff
@@ -3443,7 +3454,10 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
             } else {
                 switch (*p) {
                     case '\\':
-                        backslash = true;
+                        if (inquote || !splitEscapedBoundary || (p[1] != ';') ||
+                            !isMimeParameter(p + 2, "boundary")) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Fixed in 6c3126ac1. Canonical boundary parsing now looks past the escaped semicolon for a later boundary parameter, then recognizes the whitespace-delimited boundary while processing the intervening non-boundary argument. I also added an encrypted end-to-end fixture for the exact `boundary=X\; name=a boundary=Y` form. During regression testing I caught and corrected an over-broad first draft so the existing malformed `boundary=X boundary=Z` value remains intact. The full CTest suite and the focused libclamav/clamscan tests pass.

File: libclamav/mbox.c [hunk 10]
```diff
@@ -3426,24 +3474,34 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+
         while (*p) {
             if (backslash) {
                 backslash = false;
             } else {
                 switch (*p) {
                     case '\\':
-                        backslash = true;
+                        if (inquote || !splitBoundaryArguments || (p[1] != ';') ||
+                            !hasMimeParameterAhead(p + 2, "boundary")) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Fixed in 1679894f4. Canonical boundary parsing now ends an unquoted boundary at a following ordinary MIME parameter, so `boundary=X\; name=a` retains `X\;` rather than absorbing ` name=a`. The token search also avoids treating parameter-shaped text inside quotes as a new boundary. I added an encrypted end-to-end fixture for this exact header and verified it alongside the intervening-parameter, lone-semicolon, repeated-boundary, and space-containing-boundary fixtures. The full CTest suite passes.

File: libclamav/mbox.c [hunk 11]
```diff
@@ -3426,24 +3464,51 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
         while (*p) {
             if (backslash) {
                 backslash = false;
             } else {
                 switch (*p) {
                     case '\\':
-                        backslash = true;
+                        if (inquote || !splitBoundaryArguments || (p[1] != ';') ||
+                            !isMimeParameter(p + 2, "boundary")) {
+                            backslash = true;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3714]: Fixed in ae831fb23. Canonical boundary parsing now treats escaped whitespace followed by another MIME parameter as a delimiter while retaining the preceding backslash in the boundary value. The new end-to-end fixture covers boundary=X\ name=a and confirms the embedded attachment is scanned.

File: libclamav/message.c [hunk 12]
```diff
@@ -390,36 +390,37 @@ messageGetDispositionType(const message *m)
...
     if (!usefulArg(arg))
-        return;
+        return true;
 
     for (offset = 0; offset < m->numberOfArguments; offset++)
         if (m->mimeArguments[offset] == NULL)
             break;
         else if (strcasecmp(arg, m->mimeArguments[offset]) == 0)
-            return; /* already in there */
+            return true; /* already in there */
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/message.c]: Fixed in c8d6015f8. An exact decoded duplicate is now moved after earlier legacy values so last-value lookup keeps the canonical RFC 2231 result authoritative. The regression fixture also uses a differently cased earlier boundary value to verify that case-sensitive boundary spelling is preserved.

File: libclamav/mbox.c [hunk 13]
```diff
@@ -3452,6 +3523,16 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
                         if (!inquote)
                             goto done;
                         break;
+                    default:
+                        if (!inquote && splitBoundaryArguments && isspace((unsigned char)*p)) {
+                            bool nextIsBoundary = isMimeParameter(p, "boundary");
+
+                            if ((!argumentIsBoundary && nextIsBoundary) ||
+                                (argumentIsBoundary && !nextIsBoundary && isMimeParameterToken(p))) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Thanks for flagging this. I checked it against the original CLAM-3074 report, and splitting this form would reintroduce the reported evasion. The report's exact construction is `boundary=X boundary=Z`; Outlook, New Outlook, and Thunderbird treat the whitespace as part of the boundary, and the executable part uses `--X boundary=Z`. Our existing `clam.mail-boundary-space-delimiter.eml.xor` fixture reproduces that exact header with a decoy part under `--X` and the test attachment under `--X boundary=Z`, and it passes with the current canonical value. Treating the second `boundary=` as a separate parameter would instead select `Z`, for which the message has no delimiter. The quoted example likewise has no semicolon introducing a second parameter, so the text after the closed quoted value does not justify overriding `X` with `Y`. I am leaving this behavior unchanged.

File: libclamav/mbox.c [hunk 14]
```diff
@@ -3472,13 +3553,309 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+        quoted = true;
+        input++;
+    }
+
+    while (*input != '\0') {
+        if (quoted && (*input == '"'))
+            break;
+
+        if (quoted && (*input == '\\') && (input[1] != '\0'))
+            input++;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3861]: Confirmed and fixed in 5d2c120e6. Comment removal now preserves a backslash only for quoted strings, so unquoted quoted-pairs retain the legacy one-pass decoding behavior while quoted boundary values still reach the canonical decoder intact. I added encrypted end-to-end fixtures covering an unquoted quoted-pair, escaped semicolon, and escaped whitespace in a comment-bearing Content-Type. During the adjacent review I also fixed whitespace-before-separator and escaped-semicolon look-ahead cases and added coverage for those forms. The complete 7-target CTest suite, direct fixture scans, a 17-case standards matrix, a 50,141-message differential comparison, the long-header performance check, and make install all pass.

File: libclamav/mbox.c [hunk 15]
```diff
@@ -3426,24 +3535,62 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+                if (seekBackslash) {
+                    seekBackslash = false;
+                    if (isspace((unsigned char)*p) && isCanonicalBoundaryParameter(p))
+                        break;
+                } else if (*p == '\\') {
+                    seekBackslash = true;
+                } else if (*p == '"') {
+                    seekInquote = !seekInquote;
+                } else if (!seekInquote && isspace((unsigned char)*p) &&
+                           isCanonicalBoundaryParameter(p)) {
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in 394487b17. Both boundary look-ahead paths now locate the end of each contiguous whitespace run once, check the following token once, and then advance over or copy the complete run without changing quote, escape, delimiter, or output-buffer state. I added clamscan coverage for long folded whitespace in both the initial seek and boundary-argument paths. The focused libclamav and clamscan tests, make -j12, and git diff --check pass. I also rechecked the complete PR delta with a 41-case independent standards matrix and a 50,141-case branch-vs-main differential run; no additional regressions, crashes, or hangs were found.

File: libclamav/mbox.c [hunk 16]
```diff
@@ -3472,13 +3667,310 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+        const char *suffix;
+        char *decoded = NULL;
+        size_t nameLength;
+        size_t section = 0;
+        bool encoded   = false;
+
+        order++;
+        argumentCount++;
+        if (argumentCount >= HEURISTIC_EMAIL_MAX_ARGUMENTS_PER_HEADER)
+            break;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Thank you — I confirmed the mismatch and fixed it in 678d16eea. The canonical resolver now completes its bounded linear pass, while the legacy parser counts each whitespace-separated parameter it actually processes and raises the existing argument-limit heuristic when appropriate. I also found that simply discarding a truncated canonical result would still miss an RFC 2231 boundary split across segments, so the new regression coverage exercises direct, extended, and multi-segment boundaries at the threshold as well as the heuristic path. `make -j12`, the 137-test clamscan suite, and the full 7-test CTest suite all pass locally.

File: libclamav/mbox.c [hunk 17]
```diff
@@ -3426,24 +3535,76 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+            while ((*p != '\0') && (*p != ';')) {
+                if (seekBackslash) {
+                    seekBackslash = false;
+                    if (isspace((unsigned char)*p)) {
+                        const char *next = p;
+
+                        while (isspace((unsigned char)*next))
+                            next++;
+                        if (isCanonicalBoundaryParameter(next))
+                            break;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3629]: Confirmed and fixed in 710b0fca5. The initial media-type seek now performs escaped-whitespace boundary lookahead only outside quoted text, matching the quote-state guards already used by the later parameter scan. I added end-to-end coverage for the reported quoted media type with a fake inner boundary and a real following boundary. The case failed clean before the fix and now extracts and detects the embedded test file. `make -j12`, all 138 clamscan tests, and the complete 7-target CTest suite pass. I also reviewed the adjacent quote/escape state transitions and found no second missing quote-state guard.

File: libclamav/mbox.c [hunk 18]
```diff
@@ -3426,24 +3535,76 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
                 switch (*p) {
                     case '\\':
-                        backslash = true;
+                        if (inquote || !splitBoundaryArguments || (p[1] != ';') ||
+                            !isCanonicalBoundaryParameter(p + 2)) {
+                            backslash = true;
+                        }
                         break;
                     case '"':
                         inquote = !inquote;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in ad2375135. Canonical tokenization now opens quoted syntax only when the quote is the first non-whitespace character of a parameter value; a quote embedded in an already-started unquoted value remains data and cannot consume later fields. The regression test covers the ordinary-parameter form and the related boundary-value form, plus the prior quoted-media-type case. I used balanced malformed variants for the end-to-end cases so they isolate boundary selection rather than ClamAV’s separate odd-quote header-folding recovery. `make -j12`, all 138 clamscan tests, libclamav, and the complete seven-target suite pass; clamd/FreshClam were rerun with the local socket permissions their tests require.

File: libclamav/mbox.c [hunk 19]
```diff
@@ -3472,13 +3685,306 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+        bool encoded   = false;
+
+        order++;
+
+        while (isspace((unsigned char)*nameStart))
+            nameStart++;
+
+        separator = strchr(nameStart, '=');
+        if (separator == NULL)
+            continue;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3970]: Confirmed and fixed in dbdcbf422. Canonical boundary tokenization and resolution now both recognize ClamAV’s legacy colon separator, so a later `boundary:"Y"` remains authoritative just as it was through `messageAddArguments()`. I also checked the adjacent paths and added end-to-end coverage for direct quoted values, RFC 2231 extended and continuation forms, escaped-semicolon and whitespace-packed declarations, a missing-semicolon lookahead, and the legacy rule that `=` takes precedence when both separators appear in one argument. `make -j12`, all 139 clamscan tests, libclamav, Rust, sigtool, examples, clamd, and FreshClam pass locally.

File: libclamav/mbox.c [hunk 20]
```diff
@@ -3426,32 +3535,136 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+                            next++;
+                        if (isCanonicalBoundaryParameter(next))
+                            break;
+                        p = next;
+                        continue;
+                    }
+                } else if (*p == '\\') {
+                    seekBackslash = true;
+                } else if (*p == '"') {
+                    seekInquote = !seekInquote;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in 4f96ce139. The resume/seek loop now tracks separator, token, and value state just like the argument-copy loop, so an embedded quote cannot open quoted syntax and hide a later boundary. The end-to-end regression uses a balanced malformed value to isolate this tokenizer from ClamAV’s separate odd-quote header-folding behavior, and covers both `boundary=Y` and `boundary:"Y"`. During the adjacent review I also found that the seek loop stopped at semicolons before consulting quote state; that could promote `boundary=BAD` text inside a genuinely quoted ordinary parameter. The same commit makes semicolons delimit only outside quoted/escaped text and covers quoted colon values and quoted semicolons. `make -j12`, all 139 clamscan tests, libclamav, Rust, sigtool, examples, clamd, and FreshClam pass locally.

File: libclamav/mbox.c [hunk 21]
```diff
@@ -3426,32 +3535,164 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+                            bool nextIsBoundary;
+
+                            while (isspace((unsigned char)*next))
+                                next++;
+                            nextIsBoundary = isCanonicalBoundaryParameter(next);
+
+                            /* Before the separator, whitespace may belong to
+                             * a tolerated form such as boundary *0=value. */
+                            if (!argumentIsBoundary && nextIsBoundary)
+                                goto done;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in 5d00a028b. The compatibility exception that keeps whitespace inside an ordinary `boundary=`/`boundary:` value is now limited to that ordinary form; RFC 2231 extended and continued values split before adjacent declarations. The adjacent review also found that extended values followed by arbitrary unquoted text had the same legacy mismatch, so they now end at whitespace once the value has started while quoted spaces remain intact. End-to-end coverage includes the reported extended/direct pair, `boundary*0`/`boundary*1` pairs with both `=` and `:`, trailing text, escaped whitespace, and quoted spaces. The existing ordinary space-delimiter fixture continues to pass. `make -j12`, all 140 clamscan tests, libclamav, Rust, sigtool, examples, clamd, and FreshClam pass locally.

File: libclamav/mbox.c [hunk 22]
```diff
@@ -3426,32 +3561,174 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+
+                            while (isspace((unsigned char)*next))
+                                next++;
+                            nextIsBoundary = isCanonicalBoundaryParameter(next);
+
+                            /* Before the separator, whitespace may belong to
+                             * a tolerated form such as boundary *0=value. */
+                            if (!argumentIsOrdinaryBoundary && nextIsBoundary)
+                                goto done;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c]: Confirmed and fixed in 24fe76cad. Whitespace adjacency is now preserved only when both the current and following declarations are ordinary boundaries. If either side is RFC 2231 extended/continued, the declarations split and the later canonical value remains authoritative. I added end-to-end coverage for ordinary→extended, ordinary→continuation, and quoted ordinary→extended orderings, including a colon-separated extended value; the existing ordinary→ordinary space-delimiter fixture still passes. `make -j12`, all 140 clamscan tests, libclamav, Rust, sigtool, examples, clamd, and FreshClam pass locally.

File: libclamav/mbox.c [hunk 23]
```diff
@@ -3426,32 +3561,178 @@ nextMimeArgument(const char *ptr, char *buf, size_t buflen)
...
+                            argumentValueStarted = true;
+                        if (inquote || !splitBoundaryArguments || (p[1] != ';') ||
+                            !isCanonicalBoundaryParameter(p + 2)) {
+                            backslash = true;
+                        }
                         break;
                     case '"':
-                        inquote = !inquote;
+                        if (inquote) {
+                            inquote = false;
```
[the remark this answers is not in the document]
Comment (val-ms) [reply libclamav/mbox.c:3719]: Confirmed and fixed in d93aec429. Canonical tokenization now ends a quoted boundary when a parameter-shaped token begins immediately after the closing quote, and the resumed seek path recognizes the same construction after an intervening quoted parameter. The adjacent review also covered colon separators, ordinary and RFC 2231 extended declarations in both orders, continuations, and a quoted intervening parameter. Negative coverage preserves the intentional whitespace form (`boundary="X" boundary=Y`) and literal text (`boundary="X"foo boundary=Y`). `make -j12`, all clamscan tests, libclamav, Rust, sigtool, examples, clamd, and FreshClam pass locally; `git diff --check` is clean.

**Comment by val-ms** [comment]
@codex review

**Review by jhumlick — CHANGES_REQUESTED** [review]
libclamav/mbox.c:3492-3495 — isCanonicalBoundaryParameter() accepts whitespace inside RFC 2231 parameter names, e.g. boundary* =Y. findMimeBoundary() then trims that whitespace and treats it as a valid extended boundary, appending boundary=Y as authoritative.

Content-Type: multipart/mixed; boundary=X; boundary* =Y

With body delimiters --X, main detects the attachment, but PR 1814 selects Y and returns clean. The same occurs with boundary*0 =Y.
Please require the RFC 2231 suffix (*, *0, *0*) to be immediately adjacent to the separator, or otherwise align canonical validation with the legacy parser. Add regression coverage for these malformed forms.

**Comment by val-ms** [comment]
Thanks for catching this, @jhumlick. Fixed in 49c7471d8.

I consolidated canonical boundary-name recognition and final boundary selection around one parser. RFC 2231 suffixes now have to be immediately followed by `=` or the legacy `:` separator, so malformed forms such as `boundary* =Y`, `boundary*0 =Y`, and `boundary*0* =Y` cannot become authoritative. I also checked the adjacent forms and added end-to-end coverage for direct, continued, encoded, equals-separated, colon-separated, semicolon-delimited, and whitespace-adjacent cases.

The compatibility cases remain intact: ordinary `boundary =...` / `boundary :...` parameters and the branch's existing `boundary *0=...` handling still work, and colon/equal precedence remains aligned with the legacy parser.

Local validation passed the full build and all 7 CTest targets, including 141 clamscan tests and 15 clamd tests. Could you please take another look?

**Review by jhumlick — APPROVED** [review]
